refactor: centralize CoT parsing in backend for streaming mode

ServeurpersoCom opened the pull request:
Refactored try_parse_reasoning() to handle incremental parsing during streaming:
Parser improvements:

Tracks partial tag detection (e.g., when stream cuts mid-tag like <thi...)
Handles multiple consecutive reasoning segments within a single response
Preserves leading whitespace while detecting reasoning blocks
Continues processing content after </think> closes instead of stopping early
Works correctly with thinking_forced_open flag for grammar-constrained generation

Integration changes:

Modified common_chat_parse_content_only() and common_chat_parse_llama_3_1() to invoke reasoning parsing before tool call handling
Changed default reasoning_format from auto to deepseek for consistent behavior
Added deepseek-legacy option for backwards compatibility (inlines tags in content)

Benefits

Clients no longer need custom CoT parsing logic for streaming mode
Consistent API behavior: reasoning_content and content properly separated in both streaming and non-streaming modes
Simplifies webui and SDK implementations
Universal: works across all reasoning formats, not just DeepSeek

When generation is launched from a template that ends the system prompt with <think> and thinking is enabled, the template sets thinking_forced_open = true; the parser then consumes all text preceding </think> as reasoning, even if the opening tag is never emitted by the model, as validated by the REASONING</think>CONTENT test.

Without --jinja, the server prohibits the use of tools/tool_choice, sets inputs.use_jinja = false, and computes enable_thinking = false. Templates therefore immediately close the <think> tag and no longer force the opening, which disables the reasoning/content separation required by our new scenarios (the parser is still called on the stream but will only see regular content unless the model itself emits complete tags).

Testing
Added parser tests covering:

Inline <think>...</think> segments in CONTENT_ONLY format
Inline reasoning in LLAMA_3_X format
Multiple reasoning blocks in single response
Partial tag detection during streaming

ServeurpersoCom committed:
refactor: unify reasoning handling via backend reasoning_content, drop frontend tag parsing

- Updated the chat message component to surface backend-supplied reasoning via message.thinking while showing the raw assistant content without inline tag scrubbing
- Simplified chat streaming to append content chunks directly, stream reasoning into the message model, and persist any partial reasoning when generation stops
- Refactored the chat service SSE handler to rely on server-provided reasoning_content, removing legacy <think> parsing logic
- Refreshed Storybook data and streaming flows to populate the thinking field explicitly for static and streaming assistant messages

ServeurpersoCom committed:
refactor: implement streaming-aware universal reasoning parser

Remove the streaming mode limitation from --reasoning-format by refactoring
try_parse_reasoning() to handle incremental parsing of <think> tags across
all formats.

- Rework try_parse_reasoning() to track whitespace, partial tags, and
  multiple reasoning segments, allowing proper separation of reasoning_content
  and content in streaming mode
- Parse reasoning tags before tool call handling in content-only and Llama 3.x
  formats to ensure inline <think> blocks are captured correctly
- Change default reasoning_format from 'auto' to 'deepseek' for consistent
  behavior
- Add 'deepseek-legacy' option to preserve old inline behavior when needed
- Update CLI help and documentation to reflect streaming support
- Add parser tests for inline <think>...</think> segments

The parser now continues processing content after </think> closes instead of
stopping, enabling proper message.reasoning_content and message.content
separation in both streaming and non-streaming modes.

Fixes the issue where streaming responses would dump everything (including
post-thinking content) into reasoning_content while leaving content empty.

allozaur commented on tools/server/webui/src/lib/components/app/chat/ChatMessages/ChatMessage.svelte:
```suggestion
		messageContent={message.content}
```

allozaur commented on tools/server/webui/src/lib/stores/chat.svelte.ts:
```diff
@@ -644,18 +642,18 @@ class ChatStore {
 		if (lastMessage && lastMessage.role === 'assistant') {
 			try {
-				const partialThinking = extractPartialThinking(this.currentResponse);
+				const contentToSave = this.currentResponse;
```
```suggestion
```

I think that there's no need to create a variable just to rename this.

allozaur commented on tools/server/webui/src/lib/stores/chat.svelte.ts:
```diff
@@ -644,18 +642,18 @@ class ChatStore {
 					timings?: ChatMessageTimings;
 				} = {
-					content: partialThinking.remainingContent || this.currentResponse
+					content: contentToSave
```
```suggestion
					content: this.currentResponse
```

allozaur commented on tools/server/webui/src/lib/services/chat.ts:
```diff
@@ -256,10 +256,8 @@ export class ChatService {
 
 		const decoder = new TextDecoder();
-		let fullResponse = '';
+		let aggregatedContent = '';
```
I think we could have a more consistent naming for these variables... @ServeurpersoCom I'm open to your suggestions :)

allozaur commented on tools/server/webui/src/stories/ChatMessage.stories.svelte:
The Chat Message stories must be updated to match this change. We don't need <think> / [THINK] processing as it's not present anymore in the chat completion responses' content anymore. Instead we'd just need a story for a regular message and one with reasoning content.

ngxson commented on common/arg.cpp:
```diff
@@ -3429,8 +3429,9 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
         "- none: leaves thoughts unparsed in `message.content`\n"
-        "- deepseek: puts thoughts in `message.reasoning_content` (except in streaming mode, which behaves as `none`)\n"
-        "(default: auto)",
+        "- deepseek: puts thoughts in `message.reasoning_content`\n"
+        "- deepseek-legacy: keeps `<think>` tags in `message.content` while also populating `message.reasoning_content`\n"
+        "(default: deepseek)",
```
could you explain why this is also changed?

ServeurpersoCom committed:
refactor: address review feedback from allozaur

- Passed the assistant message content directly to ChatMessageAssistant to drop the redundant derived state in the chat message component
- Simplified chat streaming updates by removing unused partial-thinking handling and persisting partial responses straight from currentResponse
- Refreshed the ChatMessage stories to cover standard and reasoning scenarios without the old THINK-tag parsing examples

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
refactor: restore forced reasoning prefix to pass test-chat ([chat] All tests passed)

- store the exact sequence seen on input when 'thinking_forced_open' enforces a reasoning block
- inject this prefix before the first accumulated segment in 'reasoning_content', then clear it to avoid duplication
- repeat the capture on every new 'start_think' detection to properly handle partial/streaming flows

ServeurpersoCom committed:
refactor: address review feedback from ngxson

ServeurpersoCom commented:
Sure, you can squint at curl -N chunk dumps, but this integrated UI turns that pain into a proper workflow: showing the raw wire stream (no backend parsing, no frontend Markdown) so you can actually inspect behavior across models in real time, all in one click, with a tiny commit.

<img width="1214" height="987" alt="Sans titre" src="https://github.com/user-attachments/assets/ae6da92e-3dbc-4e7c-953b-6403aca122e2" />

ServeurpersoCom committed:
debug: say goodbye to curl -N, hello one-click raw stream

- adds a new checkbox in the WebUI to display raw LLM output without backend parsing or frontend Markdown rendering

tommarques56 commented on tools/server/webui/src/lib/components/app/chat/ChatMessages/ChatMessage.svelte:
```diff
@@ -165,7 +150,7 @@
 		{editedContent}
 		{isEditing}
 		{message}
```
## 🚨 CRITICAL ISSUE - Line 15

**Issue:** Security vulnerability in message content handling

**Suggestion:** Sanitize message content before rendering to prevent XSS attacks

**Code to review:**
```diff
+		messageContent={message.content}
```

**Suggested fix:**
```
messageContent={sanitizeHtml(message.content)}
```

**⚠️ This is a critical issue that requires immediate attention!**

*Automated critical review by PR Code Review Service*

ServeurpersoCom replied:
After analysis and simulation on both master and this branch, there is no XSS.

The Markdown pipeline used by MarkdownContent.svelte converts text through remark -> remarkRehype -> rehypeStringify, without ever enabling the allowDangerousHtml option. In this default configuration, raw HTML nodes from the Markdown are rendered harmless, either escaped or removed, before being returned as an HTML string, which neutralizes fragments such as <script> or <img onerror=...> provided by the model.

Once this sanitized string is obtained, the component injects it into the DOM using {@html processedHtml}. This injection step does not execute any malicious code, since any potentially dangerous content has already been filtered out by the previous Markdown processing stage.
In summary, the simulator does not observe any XSS payloads being executed because the Markdown to HTML conversion applies these safeguards by default, preventing dangerous tags or protocols from reaching the browser intact.

ServeurpersoCom committed:
Update tools/server/webui/src/lib/components/app/chat/ChatMessages/ChatMessage.svelte

Co-authored-by: Aleksander Grygier <aleksander.grygier@gmail.com>

ServeurpersoCom committed:
webui: add Storybook example for raw LLM output and scope reasoning format toggle per story

- Added a Storybook example that showcases the chat message component in raw LLM output mode with the provided trace sample
- Updated every ChatMessage story to toggle the disableReasoningFormat setting so the raw-output rendering remains scoped to its own example

ServeurpersoCom committed:
npm run format

ggerganov reviewed (approved):
I did some testing and haven't spotted any issues. Let's wait for @ngxson approval and merge.

ngxson commented on common/chat-parser.cpp:
```diff
@@ -178,28 +198,120 @@ bool common_chat_msg_parser::try_parse_reasoning(const std::string & start_think
-            }
-            auto rest = consume_rest();
+
+    if (syntax_.reasoning_format == COMMON_REASONING_FORMAT_NONE) {
+        return false;
+    }
```
I think this condition can be moved to the top of this function for visibility

ngxson commented on common/chat-parser.cpp:
```diff
@@ -178,28 +198,120 @@ bool common_chat_msg_parser::try_parse_reasoning(const std::string & start_think
             if (!rest.empty()) {
                 handle_reasoning(rest, /* closed */ !is_partial());
             }
-            // Allow unclosed thinking tags, for now (https://github.com/ggml-org/llama.cpp/issues/13812, https://github.com/ggml-org/llama.cpp/issues/13877)
-            // if (!syntax_.thinking_forced_open) {
-            //     throw common_chat_msg_partial_exception(end_think);
-            // }
```
Did we also resolve this TODO?

ServeurpersoCom replied:
https://github.com/ggml-org/llama.cpp/pull/16426/commits/6041e2562fd4a47f4aa47a0a22af12f1ea752dbb

ngxson reviewed (commented):
Looks good over all, just some small comments

ServeurpersoCom committed:
chat-parser: address review feedback from ngxson

Co-authored-by: Xuan Son Nguyen <thichthat@gmail.com>

allozaur committed:
Merge branch 'master' into streaming-aware-cpp-parser

allozaur commented:
@ServeurpersoCom I've solved one conflict that appeared after merging #16282. Let's have some final testing round, build the fresh static output and after the manual tests and CI pass, let's ship this 🚀😄
